### PR TITLE
Add type hints (Part 2)

### PR DESCRIPTION
**Pull request by desilinguist**
### Changes

- Minor typos in docstrings.
- Add `sphinx-autodoc-typehints` dependency.
- Improve sphinx auto documentation.
- Add type hints to `container.py`
- Add type hints to `analyzer.py`
- Add type hints to `comparer.py`.
- Add type hints to `configuration_parser.py`

### To review

- Please check that I have not left any parameters/arguments/return values without type hints. 
- Here's the [RTD build for this branch](https://rsmtool.readthedocs.io/en/326-type-hints-2). Check the "API Documentation" section (only for the files affected by this MR and the previous one) to confirm that all types are linked to relevant pages from external documentations (e.g., Python, Numpy, SKLL). There's no W&B yet since it doesn't seem to work. I have a ticket open with them.
- Check for typos and/or weird issues in docstrings.

### CODE REVIEW

File: rsmtool/analyzer.py [hunk 1]
```diff
@@ -1045,15 +1064,15 @@ def filter_metrics(self, df_metrics, use_scaled_predictions=False, chosen_metric
 
     def compute_metrics(
         self,
-        df,
-        compute_shortened=False,
-        use_scaled_predictions=False,
+        df: pd.DataFrame,
+        compute_shortened: bool = False,
+        use_scaled_predictions: bool = False,
         include_second_score=False,
```
Comment (tamarl08) [inline rsmtool/analyzer.py:1070]:
```suggestion
        include_second_score: bool = False,
```

File: rsmtool/comparer.py [hunk 2]
```diff
@@ -101,17 +105,14 @@ def _modify_eval_columns_to_ensure_version_compatibilty(
...
+        existing_eval_cols : List[str]
             The existing evaluation columns.
-        short_metrics_list : list
+        short_metrics_list : List[str]
             The list of columns for the short metrics file.
-        raise_warnings : bool, optional
-            Whether to raise warnings.
-            Defaults to ``True``.
 
         Returns
```
Comment (tamarl08) [inline rsmtool/comparer.py:117]: update types in return section!

File: rsmtool/comparer.py [hunk 3]
```diff
@@ -331,23 +337,26 @@ def load_rsmtool_output(self, filedir, figdir, experiment_id, prefix, groups_eva
         experiment_id : str
             Original ``experiment_id`` used to generate the output files.
         prefix: str
-            Must be set to ``scale`` or ``raw``. Indicates whether the score
+            Must be set to ``"scale"`` or ``"raw"``. Indicates whether the score
             is scaled or not.
         groups_eval: list
```
Comment (tamarl08) [inline rsmtool/comparer.py:342]:
```suggestion
        groups_eval: List[str]
```

File: rsmtool/configuration_parser.py [hunk 4]
```diff
@@ -36,7 +37,9 @@
 from .utils.models import is_skll_model
 
 
-def configure(context, config_file_or_obj_or_dict):
+def configure(
+    context: str, config_file_or_obj_or_dict: Union[str, "Configuration", Dict[str, Any], Path]
```
Comment (tamarl08) [inline rsmtool/configuration_parser.py:41]:
```suggestion
    context: str, config_file_or_obj_or_dict: Union[str, Configuration, Dict[str, Any], Path]
```

File: rsmtool/configuration_parser.py [hunk 5]
```diff
@@ -36,7 +37,9 @@
...
+) -> "Configuration":
```
Comment (tamarl08) [inline rsmtool/configuration_parser.py:42]:
```suggestion
) -> Configuration:
```

**Review by mulhod — APPROVED** [review]

File: rsmtool/configuration_parser.py [hunk 6]
```diff
@@ -319,29 +329,29 @@ def get(self, key, default=None):
...
-        keys : list of str
+        keys : List[str]
             A list of keys in the configuration object.
         """
         return [k for k in self._config.keys()]
 
-    def values(self):
+    def values(self) -> List[Any]:
         """
         Return configuration values as a list.
```
Comment (tamarl08) [inline rsmtool/configuration_parser.py:357]: update return type below.

File: rsmtool/configuration_parser.py [hunk 7]
```diff
@@ -352,44 +362,43 @@ def values(self):
...
-        items : list of tuples
-            A list of (key, value) tuples in the
-            configuration object.
+        items : List[Tuple[str, Any]]
+            A list of (key, value) tuples in the configuration object.
         """
         return [(k, v) for k, v in self._config.items()]
 
-    def pop(self, key, default=None):
+    def pop(self, key, default=None) -> Any:
```
Comment (tamarl08) [inline rsmtool/configuration_parser.py:376]:
```suggestion
    def pop(self, key: str, default: Any=None) -> Any:
```

File: rsmtool/configuration_parser.py [hunk 8]
```diff
@@ -446,7 +454,9 @@ def check_exclude_listwise(self):
             exclude_listwise = True
         return exclude_listwise
 
-    def check_flag_column(self, flag_column="flag_column", partition="unknown"):
+    def check_flag_column(
+        self, flag_column: str = "flag_column", partition: str = "unknown"
```
Comment (tamarl08) [inline rsmtool/configuration_parser.py:458]: in the doctoring it says  partition defaults to "both". which is right?

File: rsmtool/configuration_parser.py [hunk 9]
```diff
@@ -469,8 +479,7 @@ def check_flag_column(self, flag_column="flag_column", partition="unknown"):
         Returns
         -------
         new_filtering_dict : dict
```
Comment (tamarl08) [inline rsmtool/configuration_parser.py:481]:
```suggestion
        new_filtering_dict : Dict[str, List[str]]
```

File: rsmtool/container.py [hunk 10]
```diff
@@ -167,39 +178,40 @@ def to_datasets(data_container):
 
         Returns
         -------
-        datasets_dict : list of dicts
+        datasets_dict : List[DatasetDict]
```
Comment (tamarl08) [inline rsmtool/container.py:181]:
```suggestion
        datasets_dicts : List[DatasetDict]
```

File: rsmtool/container.py [hunk 11]
```diff
@@ -235,8 +237,8 @@ def get_path(self, name: str, default: Optional[str] = None) -> Optional[str]:
 
         Returns
         -------
-        path : str
-            The path for the named dataset.
+        path : Optional[str]
+            The path for the named datasOptional[str]
```
Comment (tamarl08) [inline rsmtool/container.py:241]: typo!

Comment (desilinguist) [reply rsmtool/container.py:241]: Ugh 🤦🏽

**Review by tamarl08 — APPROVED** [review]

**Pull request merged**